### ISSUE TITLE
[Tracking] The Mentor/learner Relationship Project

### ISSUE THREAD

**Issue by iHiD** [exercism/exercism#4658]
This forms a big part of the work we're doing over the next 2 months. I'm opening this as both an FYI and a tracking issue. Comments are welcome :)

## Objectives

Providing a positive learning experience is key to Exercism. At the heart of this is the relationship between mentors and learners and the mutual respect required for positive interactions. This project aims to address that and has a two key objectives:

- Increase learner's respect for mentors. Mentors are volunteering their time and therefore deserve an extra amount of implicit respect. Often they'll have mentored lots of solutions and had lots of experience on a problem. If a learner respects this up front and is open to their insight, then there is a large likelyhood that the learner will learn more.
- Increase mentor's respect and understanding of learners. By improving a mentors' understanding of who a learner is and what they want to learn, feedback can be tailored, making the whole interaction more valuable.

## Solution overview

The key method we want to use to address this is to provide profiles of learners and mentors to each other during the mentoring process. These will contain relevant information that nurtures respect and understanding between the two parties. Two quick sketches to how these might look:

![img_20190203_121746](https://user-images.githubusercontent.com/286476/52182214-cdfa3e00-27af-11e9-930b-d2bcfb36c6d4.jpg)
![img_20190203_122453](https://user-images.githubusercontent.com/286476/52182215-cdfa3e00-27af-11e9-837c-09a6575d0015.jpg)

## Practical Steps

Getting this right involves quite a lot of changes. I'll add to this list as further changes become applicable:

### Mentoring solution UI
- [x] Remove the ability to approve with no comments in the discussion (every student deserves at least a "well done") (https://github.com/exercism/website/pull/565)
- [x] Allow students to give positive feedback / thanks to a mentor (https://github.com/exercism/website/pull/574)
- [ ] Change metaphor to one-mentor-per-solution
- [ ] Add mentor profiles that talk about the mentor's interests (general and for language), their experience, rating etc.
- [ ] Have mentor profile open as modal on first time a learner visits their solution once theirs been a comment.
- [ ] Add a section for "Your mentor is ..." with a link to the modal, above conversation.
- [ ] Move mentor bio editing into the website (from website-copy)
- [ ] Allow students to edit their mentor ratings.
- [ ] Quick buttons for students to add comments on what was good bad (e.g. respectful/positive/responsive for 5* | rude/slow/inflexible for <5*). Can you add your own? (data collection for us)
- [ ] Add ability for student to report mentor to admins
- [ ] Add ability for mentor to report student to admins

### Give mentors ability to review students 
- [ ] Notify mentor when solution is completed
- [ ] Rate student (1-5?) - auto-filled to 5
- [ ] Quick buttons for to add comments on what was good bad (e.g. respectful/positive/responsive for 5* | rude/slow/did-not-listen for <5*). Can you add your own? (data collection for us)

### Mentor Rank changes
- [ ] Only appear on mentors list after 10 solutions
- [ ] Only appear on mentors list if rating > 4.5 (?)
- [ ] Invited to GitHub Mentor Org after 10 solutions.
- [ ] Badge on profile after 10 solutions
- [ ] Create badge (like the ones people use on READMEs on GitHub) for mentor rating/num-solutions.

### Other tweaks 
- [ ] Exclude last 5 reviews when calculating mentor score
- [ ] Students with low ratings will be deprioritised in the queues.
- [ ] Add notifications/section for admins to deal with CoC reports
- [ ] Add "tell your mentors why you're doing this track" on "my track page" to populate the student modal.

### Still to work out
- How to mentor profiles appear away from the solution (e.g. on normal profiles)
- Actual designs for the modals, and the data needed to power this.

**Comment by kotp** [comment]
- Add ability for student to report mentor to admins
- Add ability for mentor to report mentor to admins

It looks like students can report a mentor to admins, and on next line, a mentor can report themselves to admins as well.

**Comment by marvinrabe** [comment]
I disagree with "Rate student (1-5?)". I cannot see the benefit of this. Sure, every student is different and there are some problematic ones. But most of the time they are friendly and are interested in learning more. I don't differentiate mentoring for "nice" and "evil" students. Everyone will get a review and if they are willing to dive deeper into a topic, I am glad to provide more insight. And if I there are problems with a student, I will discuss it with them in the comments in a productive way.

My fear is that in the end the mentoring dashboard will be full of "evil" students because mentors will prefer the easy ones. Or students that had some bad days will forever marked as "evil" even if they have changed and won't get the review they deserve.

**Comment by kotp** [comment]
We all know that all students are a 5 unless otherwise proven otherwise, though some may think that all students are a 3 unless they show exceptional traits either way (whatever those are).  I rate would rate all mentors as 3, as all of them are average, except for the above and below average ones, of course.

**Comment by iHiD** [comment]
@marvinrabe I think students that are respectful and nice to mentors should be treated by our systems in a more prioritised fashion than students that are rude and/or unresponsive. I can't speak to your specific experiences, by there are many times that mentors just give up on the process because of how rude/disrespectful students can be. 

For clarity, I'm not necessarily considering exposing a student's score to a mentor, but I want to be able to use it as a red flag for us, or as a prioritisation method etc (similar to how Uber uses passenger ratings). So:

> the mentoring dashboard will be full of "evil" students

I see this as unlikely as mentors won't know who is good or evil, and in fact, I'd expect the opposite - that all the people at the top of the dashboard each time you return are positive students. 

> Or students that had some bad days will forever marked as "evil" even if they have changed and won't get the mentoring they deserve.

I think for a student to get a low (1/2*) rating probably needs more than an "off day" - ie they need to be actively rude. But by taking averages and removing the top/bottom % of ratings (as we do with mentor ratings) helps avoid this problem. `9x 5*` + `1x 1*` would still be a healthy rating (`>4`) and therefore I wouldn't expect someone to be penalised for it, etc.

> I rate would rate all mentors as 3, as all of them are average, except for the above and below average ones, of course.

- ~95% of mentor ratings are 5*
- We'll prepopluate the student rating as 5*, meaning a mentor will actively have to choose to change  that value to downgrade someone (this is also what Uber does for drives)

**Comment by kotp** [comment]
Thanks @ihid. That says a lot to me, regarding the 95%.

**Comment by SaschaMann** [comment]
I feel like the 1-5 rating system, for both mentors and students, needs more explanation on what the numbers actually mean or it'll end up like this:

![](https://imgs.xkcd.com/comics/star_ratings.png)
(https://xkcd.com/1098/)

A few options to prevent that could be calculating it directly from those quick buttons, reducing that to a simple choice "I would (not) want to mentor them again" or giving specific meanings to each of the numbers (e.g. "5* - They taught me something beyond the initial scope of the exercise" or sth. along those lines).

> ~95% of mentor ratings are 5*

This makes the rating somewhat meaningless because it's nearly impossible to tell apart outstanding or above/below average mentors when all the ratings are that close together. I'm curious though, are the remaining 5% mostly 1 star ratings or somewhere inbetween?

**Comment by marvinrabe** [comment]
I agree to @SaschaMann. Probably 95% of student ratings will be 5* as well. So is the problem of rude students solved with a ranking system? Probably not.

Maybe a "Mentor reports User" will solve this. In most cases we can solve the issues ourselves. And when it becomes out of hand you see the reportings and can respond to them accordingly (e.g. banning the user, …).

**Comment by tehsphinx** [comment]
@marvinrabe: Since I have already seen really rude students (not to myself but to other mentors) I think there should be a way for mentors to let admins know there is a problem. And that without having to actively write to them on Slack or other way. Especially since as a mentor I would feel like "I might have said something wrong" or |maybe the student just had a bad day" and "this might be a one time thing". Talking to an admin about it could trigger a disproportionate response. But one bad rating wouldn't trigger any red flags. If however others have the same experience with a student, admins should know and be able to intervene. That can best be done with student ratings that accumulate to bad ratings over several exercises.

We have mentor ratings on exercism. It is ok to rate mentors and give them a bad rating -- deserved or not. But exercism lives on it's mentors. They are the ones that need protection, not the students. If a mentor quits because another student had a bad day, that is a huge loss.

So to be honest I cannot really understand the critique on the student ratings. They are an objective way for admins to see if a student needs to be talked to or even be banned. And unfortunately we live in a world where that is necessary.

**Comment by marvinrabe** [comment]
@tehsphinx I agree. But I don't think the best solution for this is a 5 star rating system. A reporting (like the mentor reporting) is probably a better one.

**Comment by rpottsoh** [comment]
Am I rating a student based on  their behavior or my opinion of their skill?  If the later could this be used to get a sense of if a student is improving?   Is the rating of a student systemwide, only the track, or only the exercise.....  several ways I can see pivoting, I've not listed them all.  I think it should be used as a positive feedback tool and not for black marking.  Behavioral issues should just be handled quietly, discretely and on case by case basis.

**Comment by iHiD** [comment]
> @tehsphinx I agree. But I don't think the best solution for this is a 5 star rating system. A reporting (like the mentor reporting) is probably a better one.

**Comment by iHiD** [comment]
> @tehsphinx I agree. But I don't think the best solution for this is a 5 star rating system. A reporting (like the mentor reporting) is probably a better one.

The problem is, reporting is a qualitative solution whereas ratings are a quantitative solution. So for reporting I have to spend 15mins every time someone reports something dealing with it, then spending time working out a solution (which is basically ban someone, argue with someone, or disagree with the mentor's assessment). A rating system leaves that whole thing to algorithms. If someone is consistently "evil" then they just get dropped to the bottom of the queue, and maybe eventually auto-banned. And if someone's behaviour improves the ratings can solve that. Ratings allow us to solve these problems algorithmically, rather than using human time.

There seem to be two different sub-debates here:
1) Should we have ratings at all?
2) Should ratings be 1-5 or some other system that is less subjective.

@SaschaMann's post is totally true, but that's not necessarily bad. In the example of Uber, if you drop below 4.5* then you get "retrained". If an unusual amount of people actively chooses to downgrade you then that's worth looking in to. Giving people the ability to vary that downgrade from 1-4 isn't necessarily bad either. And it's easy to offset an individual mentor's rating from the mean (e.g. if they rate 4* 90% and only 5* 1% of the time, then we consider their 4 someone else's 4.9*).

When discussing (2), it seems to me that the most important thing is that it's clear to the mentor what they're rating. But even without that I think people do get it in reality. The student's ratings of mentors so far are: `{1: 123, 2: 140, 3: 681, 4: 2874, 5: 25901}`, which is about what I'd expect for people treating 5* as "normal".

**Comment by lbsa71** [comment]
I have suggested on github and in slack to get rid of the 'quantitative' rating, and rather introduce a functional selection: Make those buttons mean something concrete for both the student and the mentor, as tooling to help shape our journeys together. I would say that at any point in an interaction, both students and mentors should be able to do ay of the following: "I would like to report bad behaviour to admin", "I would like to never interact with this person again", "I can interact with this person again", "I would like to prioritize interacting with this person". That is what would make my mentoring experience enjoyable. Bios and stuff, secondary for me. Ratings - detrimental.

**Comment by ErikSchierboom** [comment]
Just a random thought: would it make sense for a student to say: "I'd like to retain this mentor as my mentor?" That way, a mentor could build up more of a "connection" with a student. There are of course problems with this approach, with certain mentors perhaps being overwhelmed and such. But consider it some food for thought :)

**Comment by lbsa71** [comment]
@ErikSchierboom that was somewhat what I was aiming for with the 'Prioritize' option, the idea being that if two individuals chose to 'prioritize' each others, the system would more or less guarantee that they would cooperate through the track. But I think a priority system would allow for a degree of flexibility to cope with overwhelming, and students/mentors being temporarily less available, et c.

**Comment by ErikSchierboom** [comment]
Ah yes, that sounds like we're talking about the same thing.

**Comment by iHiD** [comment]
> Just a random thought: would it make sense for a student to say: "I'd like to retain this mentor as my mentor?" That way, a mentor could build up more of a "connection" with a student. There are of course problems with this approach, with certain mentors perhaps being overwhelmed and such. But consider it some food for thought :)

So I always like this suggestion, but have no idea how it would work in practice. For example, if we  only show the solution in the queue for the specific matched mentor but that mentor doesn't come online for a week, then the learner has a massive wait. Or we reserve it for 24hrs, but then that might still mean a slower response time in some tracks. If anyone can suggest a full solution to this, I'd be happy to look into it.

**Comment by F3PiX** [comment]
> Providing a positive learning experience is key to Exercism. 
And it's one of the greatest accomplishments as well. 👏  
I 100% agree that creating a safe and respectful environment is invaluable, and I'm aware of the effort that takes.
However, I don't see how grading students is helping with that. I'd rather expect it would work against it. 

My second objection is that  personal interpretations of what 'rude' is are going to affect the learners' experience, while the source of the behaviour could be a cultural thing or not be proficient in English, or just not knowing what Exercism's mentoring entails / expectations management.     
  
I also strongly agree with @Lirt that 'rating students' is contradictory to the 'informal learning' environment that Exercism is.  

Then there's the thing that 'flagging' a student currently means that you, Jeremy, needs to address. That is hard to scale, and the more Exercise grows, the more flags there are to be handled, and it seems that growing communities also get ruder. I'd be interested to hear how often this happens. 
 
It looks to me that  'handling flags' is the bigger problem that needs a solution. Maybe we can take  a different direction, for instance with 'Mentor Ambassadors', who act as gatekeepers. 

TLDR to answer your questions:
>* Should we have ratings at all?

I'd vote wholeheartedly not.

> Should ratings be 1-5 or some other system that is less subjective.
Not ratings by mentors, but it's clear to me that the 'flagging' problem needs a solution.

**Comment by mikedamay** [comment]
Can we include in the student (and, maybe, mentor) profile their self assessment of  what (natural) languages they speak and to what level of competency.  This would help avoid miscommunications in mentoring.

**Comment by iHiD** [comment]
@mikedamay Yes. This will be part of "Add "tell your mentors why you're doing this track" on "my track page" to populate the student modal."

**Comment by dheater** [comment]
For ranking, I like the trend toward a binary system rating systems where you select between :smile: and :frowning_face: 

That could be followed up by by a quantitative (1-5) survey sent to a small portion of student respondents about what worked or didn't during the interaction:
- response time
- usefulness of mentor comments
- problem difficulty
- instructions provided for the problem
- clarity/usefulness of test cases

Just my opinion.

**Comment by iHiD** [comment]
This will form part of v3. You can read more about the details [here](https://github.com/exercism/v3/issues/690) and open new issues in that repo to discuss things as appropriate. Thanks to everyone for the helpful conversation. I'm going to lock this now but keep it open as a reference for our product design phase of v3.

**Comment by ErikSchierboom** [comment]
Hello 👋

With the launch of Exercism v3, we are **closing all issues in this repository** to help give us a clean slate to detect new problems. If this issue is still relevant to Exercism v3 (e.g. it's a feature that we haven't implemented in v3, or a bug that still exists), please reopen it and we will review it and post an update on it as soon as we get chance.

Thanks for helping make Exercism better, and we hope you enjoy v3 🙂
